### PR TITLE
feat: write data length to first compact share

### PR DESCRIPTION
**Pull request by rootulp**
Closes https://github.com/celestiaorg/celestia-app/issues/725

### CODE REVIEW

**Comment by evan-forbes** [comment]
thanks for posting even tho its a WIP!

**Comment by rootulp** [comment]
I'm encountering a strange off by one error now. The last byte of the transaction isn't in the tx range

```
$ go test .
--- FAIL: TestTxSharePosition (0.00s)
    proof_test.go:121:
        	Error Trace:	/Users/rootulp/git/rootulp/celestia/celestia-app/pkg/prove/proof_test.go:121
        	Error:      	"j0\xb3\xbb\x7f\x8e\xfc\xaeND\xe2x\x10\xe2Wz\x8brm\x91+R\xeb\x1e\xc5[\x8c~>\u008d\x9dc\x1b¢lu]\x14\xca4\xd8K\xd7\x1a\xb9\xcfܭ)\x9a\xd29YxsÒ\xd8cB$!\x02b\x97Z \xba\xe9\xfd\xa3\xd9I\xb8\xef\xab\x01\xd5Rm\x06R%\x9a\xb0\x91Ti}\xbdr\x00I\xbd\xcf\xf4\xaf2:&\xb3U\x17\xf1\x1d\xd4>\xe3\xfb%\xec\xce\xc2\xca\xc0\xac\x14\xba\xb6\xe2\xaf:ȫ\f\xbcR\xd2\xc0WQ\f?\x9e\x14Tiny\xbc54\x8e@\xb9\x89\xc8\xdfȟ\xf9\x86Xk\xea\xf3\x12\xfc\x02\xa7\x1f\x80\r`}\xb5\x80\xc5we9(\x86G*\a\xec`<\xdf\xddS\xd0=5f\xbf\xcbK\x95\"\xb3\xd6O\x98G\x0e\x8f\x1dg\x02;`ՉߌR\x98\x85\x8fXdW\xdat\x18\xdc@\x9cG\xf6y\xd9\xeaDZײ\xa2\xf7\xa9\xebm\xbf\x83t:\xbbtiv\xcd6" does not contain "5f\xbf\xcbK\x95\"\xb3\xd6O\x98G\x0e\x8f\x1dg\x02;`ՉߌR\x98\x85\x8fXdW\xdat\x18\xdc@\x9cG\xf6y\xd9\xeaDZײ\xa2\xf7\xa9\xebm\xbf\x83t:\xbbtiv\xcd6\xa8"
        	Test:       	TestTxSharePosition
        	Messages:   	many small tx%!(EXTRA prove.startEndPoints={44 44}, int=61)
FAIL
FAIL	github.com/celestiaorg/celestia-app/pkg/prove	1.378s
FAIL
```

**Comment by rootulp** [comment]
More logging:

```
rawTx: [31 237 191 8 10 97 179 202 252 23 133 206 138 165 147 154 45 65 11 64 222 33 202 131 196 25 175 251 69 64 254 233 203 216 117 218 173 37 145 219 161 151 230 113 175 31 59 246 52 178 12 10 193 139 30 19 40 186 162 52 7 76 85 106 96 58 0 225 5 69 240 57 253 76 175 51 28 227 239 145 70 151 213 58 37 121 118 14]
rawTxDataForRange: [7 32 208 225 100 249 176 168 55 171 122 205 161 62 52 4 60 147 0 87 228 59 13 86 237 167 22 223 111 14 137 141 40 246 33 84 33 45 46 199 20 30 40 83 217 13 237 161 73 245 200 72 42 87 4 156 47 111 118 3 196 173 113 179 2 15 101 67 203 159 5 237 108 66 179 185 82 165 115 78 233 157 140 94 57 38 172 255 29 128 127 243 220 114 59 231 24 205 101 95 116 93 194 201 53 57 177 205 117 88 159 35 244 29 132 125 60 238 84 222 54 136 148 124 211 101 65 126 95 43 97 230 195 23 219 218 232 161 190 43 23 43 37 99 69 234 47 185 62 158 26 113 58 44 247 68 51 207 52 88 31 237 191 8 10 97 179 202 252 23 133 206 138 165 147 154 45 65 11 64 222 33 202 131 196 25 175 251 69 64 254 233 203 216 117 218 173 37 145 219 161 151 230 113 175 31 59 246 52 178 12 10 193 139 30 19 40 186 162 52 7 76 85 106 96 58 0 225 5 69 240 57 253 76 175 51 28 227 239 145 70 151 213 58 37 121]
```

the two missing bytes are actually in the next share. This test claims `prove.startEndPoints={81 81}` but it should have been `prove.startEndPoints={81 82}`. Either of these are incorrect and I suspect it is 1: 

1. the txSharePosition calculation
2. splitting the tx data into shares

**Comment by rootulp** [comment]
The txSharePosition calculation looks correct based on https://github.com/celestiaorg/celestia-app/pull/781/commits/bbcfd1677050f3700f918c021f78e2d37220b2cb

**Comment by rootulp** [comment]
Confirmed a bug in transaction splitting.

**Comment by evan-forbes** [comment]
If its just the proving code that is failing, then I'm ok w/ commenting those tests out for now, as this is blocking the NID implementation of tx inclusion proofs, and we'll have to rewrite/handle those bugs there anywho

**Comment by rootulp** [comment]
There is a bug in transaction splitting logic, will try to split this PR into refactors that can be merged ASAP but this PR isn't ready as is b/c there are a lot of edge cases that I don't think are handled correctly.

File: pkg/shares/compact_shares_test.go [hunk 1]
```diff
@@ -81,7 +81,7 @@ func Test_processCompactShares(t *testing.T) {
 		{"single big tx", appconsts.ContinuationCompactShareContentSize * 4, 1},
 		{"many big txs", appconsts.ContinuationCompactShareContentSize * 4, 10},
 		{"single exact size tx", exactTxShareSize, 1},
-		{"many exact size txs", exactTxShareSize, 10},
+		{"many exact size txs", exactTxShareSize, 100},
```
Comment (rootulp) [inline pkg/shares/compact_shares_test.go:84]: [for reviewers] this surfaced a bug in the implementation of this PR that a smaller txSize didn't surface so I opted to increase the number of transactions in this test

File: pkg/shares/compact_shares_test.go [hunk 2]
```diff
@@ -62,7 +62,7 @@ func TestFuzz_processCompactShares(t *testing.T) {
 
 func Test_processCompactShares(t *testing.T) {
 	// exactTxShareSize is the length of tx that will fit exactly into a single
-	// share, accounting for namespace id and the length delimiter prepended to
+	// share, accounting for the tx length delimiter prepended to
```
Comment (evan-forbes) [inline pkg/shares/compact_shares_test.go:65]: [question]
does it no longer account for the namespace as well?

File: pkg/shares/parse_compact_shares.go [hunk 3]
```diff
@@ -44,7 +45,7 @@ func (ss *shareStack) resolve() ([][]byte, error) {
 	if !infoByte.IsMessageStart() {
 		return nil, errors.New("first share is not a message start")
 	}
-	err = ss.peel(ss.shares[0][appconsts.NamespaceSize+appconsts.ShareInfoBytes+appconsts.CompactShareReservedBytes:], true)
+	err = ss.peel(ss.shares[0][appconsts.NamespaceSize+appconsts.ShareInfoBytes+appconsts.FirstCompactShareDataLengthBytes+appconsts.CompactShareReservedBytes:], true)
```
Comment (evan-forbes) [inline pkg/shares/parse_compact_shares.go:48]: nice! way to grok this and find the minimal change necessary

File: pkg/shares/split_compact_shares.go [hunk 4]
```diff
@@ -59,7 +61,7 @@ func (css *CompactShareSplitter) WriteEvidence(evd coretypes.Evidence) error {
 func (css *CompactShareSplitter) WriteBytes(rawData []byte) {
 	// if this is the first time writing to a pending share, we must add the
 	// reserved bytes
-	if len(css.pendingShare.Share) == appconsts.NamespaceSize+appconsts.ShareInfoBytes {
+	if css.isEmptyPendingShare() {
```
Comment (evan-forbes) [inline pkg/shares/split_compact_shares.go:64]: omg how did we live w/o this helper method before?

File: pkg/shares/split_compact_shares.go [hunk 5]
```diff
@@ -130,14 +132,28 @@ func (css *CompactShareSplitter) stackPending() {
...
+
+	dataLengthVarint := css.dataLengthVarint(bytesOfPadding)
+	css.writeDataLengthVarintToFirstShare(dataLengthVarint)
+	css.forceLastShareReserveByteToZero()
+	return css.shares
+}
+
+// forceLastShareReserveByteToZero overwrites the reserve byte of the last share
+// with zero. See https://github.com/celestiaorg/celestia-app/issues/779
+func (css *CompactShareSplitter) forceLastShareReserveByteToZero() {
```
Comment (evan-forbes) [inline pkg/shares/split_compact_shares.go:154]: really good thinking abstracting this

**Review by evan-forbes — APPROVED** [review]
mostly compliments and only a single doc related question.

**very** well done imo.

File: pkg/shares/compact_shares_test.go (hunk 2 above)
Comment (rootulp) [reply pkg/shares/compact_shares_test.go:65]: this comment rename could've been extracted to a separate PR but the reason I removed namespace id from the comment is because `ContinuationCompactShareContentSize` already accounts for subtracting the namespace bytes from the share size.

https://github.com/celestiaorg/celestia-app/blob/395e26819f87d8521e6ef587bbb33905f5e5da08/pkg/appconsts/appconsts.go#L35-L39

The `- 1` is an attempt at accounting for the tx length delimiter prepended to each tx.

File: pkg/shares/split_compact_shares.go (hunk 5 above)
Comment (rootulp) [reply pkg/shares/split_compact_shares.go:154]: hopefully we can remove entirely when we tackle https://github.com/celestiaorg/celestia-app/issues/779

**Comment by rootulp** [comment]
Need to merge in `main`, resolve merge conflicts, get re-approval, then this should be ready to merge

File: pkg/shares/split_compact_shares.go (hunk 5 above)
Comment (evan-forbes) [reply pkg/shares/split_compact_shares.go:154]: yeah, come to think of it, we should finish #779 before the next upgrade as well, since it is also consensus breaking

File: pkg/shares/share_splitting_test.go [hunk 6]
```diff
@@ -26,11 +26,12 @@ func TestSplitTxs(t *testing.T) {
 			want: [][]uint8{
 				append([]uint8{
 					0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x1, // namespace id
-					0x1, // info byte
-					0x0, // reserved byte
+					0x1,                // info byte
+					0x2, 0x0, 0x0, 0x0, // 1 byte (unit)  + 1 byte (unit length) = 2 bytes message length
+					0x0, // BUG: reserved byte should be non-zero
```
Comment (rootulp) [inline pkg/shares/share_splitting_test.go:31]: Opens https://github.com/celestiaorg/celestia-app/issues/817

**Review by evan-forbes — APPROVED** [review]

File: pkg/prove/proof_test.go [hunk 7]
```diff
@@ -87,6 +87,10 @@ func TestTxSharePosition(t *testing.T) {
...
+		},
```
Comment (evan-forbes) [inline pkg/prove/proof_test.go:93]: This is my bad for missing this during the review, as it looked like a single line, but why was this added? keeping this in the same line seems kinda insane given the size. If we want to keep it, we should put it in a separate test data file or something

cc @rootulp

Comment (rootulp) [reply pkg/prove/proof_test.go:93]: I'd like to keep this test case b/c it makes a previously sporadic test failure (due to randomness) a concrete test failure.

Agreed, it's a massive line. Will extract to a separate file
